modules/core/04-channel/types/upgrade.go

```diff
@@ -69,3 +72,22 @@ func (uf UpgradeFields) ValidateBasic() error {
 func (ut UpgradeTimeout) IsValid() bool {
 	return !ut.Height.IsZero() || ut.Timestamp != 0
 }
+
+// HasPassed returns true if the upgrade has passed the timeout height or timestamp
+func (ut UpgradeTimeout) HasPassed(ctx sdk.Context) (bool, string) {
+	if !ut.IsValid() {
+		return true, "upgrade timeout cannot be empty"
+	}
+
+	selfHeight, timeoutHeight := clienttypes.GetSelfHeight(ctx), ut.Height
+	if selfHeight.GTE(timeoutHeight) && timeoutHeight.GT(clienttypes.ZeroHeight()) {
+		return true, fmt.Sprintf("block height >= upgrade timeout height (%s >= %s)", selfHeight, timeoutHeight)
+	}
+
+	selfTime, timeoutTimestamp := uint64(ctx.BlockTime().UnixNano()), ut.Timestamp
+	if selfTime >= timeoutTimestamp && timeoutTimestamp > 0 {
+		return true, fmt.Sprintf("block timestamp >= upgrade timeout timestamp (%s >= %s)", ctx.BlockTime(), time.Unix(0, int64(timeoutTimestamp)))
+	}
+
+	return false, ""
```

Comment:
In the false case, we should still return a string which contains information about the timeout not being past. Maybe we should be collecting up the hasPassed, info in each if statement:

```go
var (
    hasPassed bool
    info string
)

if !timeoutHeight.IsZero() {
    hasPassed = selfHeight.GTE(timeoutHeight)
     info = fmt.Sprintf("block height %s, timeout height %s", selfHeight, timeoutHeight)
}

if timestamp != 0 {
    hasPassed = hasPassed || selfTime >= timeoutTimestamp
    info = fmt.Sprintf("%s: block timestamp %s, timeout timestamp %s", info, selfTime, timeoutTimestamp)
}

return hasPassed, info
```

Or something similar to this. I'm not sure exactly how we should format the information returned